### PR TITLE
[wasm] Source Phase Imports for ESM Integration

### PR DESCRIPTION
**Pull request by guybedford**
This adds a test for the new [source phase imports](https://github.com/tc39/proposal-source-phase-imports) aspect of the ESM integration.

Includes tests both for using the source phase, and for checking that the prototype behaviours match the updated requirements on `WebAssembly.Module` extending `AbstractModuleSource`.

Per the note in https://github.com/WebAssembly/esm-integration#es-module-integration-proposal-for-webassembly, it is possible for implementations to implement source phase imports independently from non-source-phase imports. I don't know if there's a way we want to reflect that in the test structure though?

### CODE REVIEW

**Review by lucacasonato — COMMENTED** [review]
This can be rewritten as a `.any.js` file which will make this also runnable from Web Workers and Service Workers.

**Comment by Ms2ger** [comment]
Can it? Note `<script type=module>`. The docs seem to show that the worker scopes have a module variant, but window doesn't.

**Comment by guybedford** [comment]
I've split out the tests to separate the module source constructor tests from the HTML ones, so they should work across different contexts. I've also extended the ESM integration HTML worker test to include source phase imports. PTAL

File: wasm/jsapi/module/moduleSource.any.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+// META: script=/wasm/jsapi/assertions.js
+
+let emptyModuleBinary;
+setup(() => {
+  emptyModuleBinary = new WasmModuleBuilder().toBuffer();
+});
+
+test(() => {
+  assert_equals(typeof AbstractModuleSource, "undefined");
+  const AbstractModuleSource = Object.getPrototypeOf(WebAssembly.Module).constructor;
```
Comment (nicolo-ribaudo) [inline wasm/jsapi/module/moduleSource.any.js:12]: Isn't `AbstractModuleSource` itself `Object.getPrototypeOf(WebAssembly.Module)` itself, and `.constructor` is the `Function` constructor?

Or at least, that's what I see when doing `Object.getPrototypeOf(class X extends class Y {} {})`.

```suggestion
  const AbstractModuleSource = Object.getPrototypeOf(WebAssembly.Module);
```

File: wasm/jsapi/module/moduleSource.any.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  assert_equals(AbstractModuleSource.name, "AbstractModuleSource");
+  assert_not_equals(AbstractModuleSource, Function);
+}, "AbstractModuleSource intrinsic");
+
+test(() => {
+  const AbstractModuleSourceProto = Object.getPrototypeOf(WebAssembly.Module.prototype).constructor;
```
Comment (nicolo-ribaudo) [inline wasm/jsapi/module/moduleSource.any.js:18]: `WebAssembly.Module.prototype` is a plain object, this should probably be
```suggestion
  const AbstractModuleSourceProto = Object.getPrototypeOf(WebAssembly.Module).prototype;
```

but I'm not sure about what this test is testing actually.

File: wasm/jsapi/module/moduleSource.any.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  builder
+    .addFunction("fn", kSig_v_v)
+    .addBody([])
+    .exportFunc();
+  builder.addMemory(0, 256, true);
+
+  const buffer = builder.toBuffer()
+  const module = new WebAssembly.Module(buffer);
+
+  const AbstractModuleSource = Object.getPrototypeOf(WebAssembly.Module).constructor;
```
Comment (nicolo-ribaudo) [inline wasm/jsapi/module/moduleSource.any.js:34]:
```suggestion
  const AbstractModuleSource = Object.getPrototypeOf(WebAssembly.Module);
```

File: wasm/jsapi/module/moduleSource.any.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  const toStringTag = Object.getOwnPropertyDescriptor(AbstractModuleSource.prototype, Symbol.toStringTag).get;
+  assert_equals(toStringTag.call(module), "WebAssembly.Module");
+}, "AbstractModuleSourceProto toStringTag brand check");
```
Comment (nicolo-ribaudo) [inline wasm/jsapi/module/moduleSource.any.js:37]: Can we also tests that it throws on other objects?

File: wasm/webapi/esm-integration/resources/worker.js [hunk 5]
```diff
@@ -1 +1,5 @@
-import * as mod from "./worker.wasm"
+import * as mod from "./worker.wasm";
+import source modSource from "./worker.wasm";
+assert_true(modSource instanceof WebAssembly.Module);
+assert_true(await import("./worker.wasm") === mod);
+assert_true(await import.source("./worker.wasm") === modSource);
```
Comment (nicolo-ribaudo) [inline wasm/webapi/esm-integration/resources/worker.js:5]: Lets split this in two files/tests, for implementations that support only one and not the other.

File: wasm/webapi/esm-integration/source-phase.tentative.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script type=module>
+setup({ single_test: true });
+
+import source exportedNamesSource from "./resources/exported-names.wasm";
+
+assert_true(exportedNamesSource instanceof WebAssembly.Module);
+const AbstractModuleSource = Object.getPrototypeOf(WebAssembly.Module).constructor;
```
Comment (nicolo-ribaudo) [inline wasm/webapi/esm-integration/source-phase.tentative.html:12]:
```suggestion
const AbstractModuleSource = Object.getPrototypeOf(WebAssembly.Module);
```

File: wasm/jsapi/module/moduleSource.any.js (hunk 1 above)
Comment (guybedford) [reply wasm/jsapi/module/moduleSource.any.js:12]: Yes, you're right, thank you.

File: wasm/jsapi/module/moduleSource.any.js (hunk 2 above)
Comment (guybedford) [reply wasm/jsapi/module/moduleSource.any.js:18]: Based on the feedback in the previous comment I believe this should be:

```suggestion
  const AbstractModuleSourceProto = Object.getPrototypeOf(WebAssembly.Module.prototype);
```

where what is being tested is that `WebAssembly.Module.prototype` extends from the abstract module source prototype, just as WebAssembly.Module extends from abstract module source.

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by Ms2ger — APPROVED** [review]

**Pull request merged**